src/ioctls/vm.rs

```diff
@@ -1258,62 +1284,76 @@ mod tests {
         let evtfd3 = EventFd::new(EFD_NONBLOCK).unwrap();
-        if cfg!(any(target_arch = "x86", target_arch = "x86_64")) {
-            assert!(vm_fd.create_irq_chip().is_ok());
-            assert!(vm_fd.register_irqfd(&evtfd1, 4).is_ok());
-            assert!(vm_fd.register_irqfd(&evtfd2, 8).is_ok());
-            assert!(vm_fd.register_irqfd(&evtfd3, 4).is_ok());
-        }
 
-        // On aarch64, this fails because setting up the interrupt controller is mandatory before
-        // registering any IRQ.
+        assert!(vm_fd.create_irq_chip().is_ok());
+
+        assert!(vm_fd.register_irqfd(&evtfd1, 4).is_ok());
+        assert!(vm_fd.register_irqfd(&evtfd2, 8).is_ok());
+        assert!(vm_fd.register_irqfd(&evtfd3, 4).is_ok());
+        assert!(vm_fd.unregister_irqfd(&evtfd2, 8).is_ok());
+        // KVM irqfd doesn't report failure on this case:(
+        assert!(vm_fd.unregister_irqfd(&evtfd2, 8).is_ok());
+
+        // Duplicated eventfd registration.
         // On x86_64 this fails as the event fd was already matched with a GSI.
         assert!(vm_fd.register_irqfd(&evtfd3, 4).is_err());
         assert!(vm_fd.register_irqfd(&evtfd3, 5).is_err());
+        // KVM irqfd doesn't report failure on this case:(
+        assert!(vm_fd.unregister_irqfd(&evtfd3, 5).is_ok());
     }
 
     #[test]
-    #[cfg(any(
-        target_arch = "x86",
-        target_arch = "x86_64",
-        target_arch = "arm",
-        target_arch = "aarch64"
-    ))]
-    fn test_unregister_irqfd() {
+    #[cfg(target_arch = "aarch64")]
+    fn test_register_unregister_irqfd() {
         let kvm = Kvm::new().unwrap();
         let vm_fd = kvm.create_vm().unwrap();
         let evtfd1 = EventFd::new(EFD_NONBLOCK).unwrap();
         let evtfd2 = EventFd::new(EFD_NONBLOCK).unwrap();
         let evtfd3 = EventFd::new(EFD_NONBLOCK).unwrap();
-        if cfg!(any(target_arch = "x86", target_arch = "x86_64")) {
-            assert!(vm_fd.create_irq_chip().is_ok());
-            assert!(vm_fd.register_irqfd(&evtfd1, 4).is_ok());
-            assert!(vm_fd.register_irqfd(&evtfd2, 8).is_ok());
-            assert!(vm_fd.register_irqfd(&evtfd3, 4).is_ok());
-            assert!(vm_fd.unregister_irqfd(&evtfd2, 8).is_ok());
-            // KVM irqfd doesn't report failure on this case:(
-            assert!(vm_fd.unregister_irqfd(&evtfd2, 8).is_ok());
-        }
+
+        // Create the vGIC device.
+        let vgic_fd = create_gic_device(&vm_fd, 0);
+
+        // refer: https://github.com/firecracker-microvm/firecracker/blob/master/src/arch/src/aarch64/layout.rs
+        // Dummy interrupt for testing on aarch64.
+        let nr_irqs: u32 = 128;
+
+        // We need to tell the kernel how many irqs to support with this vgic.
+        let vgic_attr = kvm_bindings::kvm_device_attr {
+            group: kvm_bindings::KVM_DEV_ARM_VGIC_GRP_NR_IRQS,
+            attr: 0,
+            addr: &nr_irqs as *const u32 as u64,
```

Comment:
There's no need for all the conversions. You can just write `addr: 128` or whatever else number you want to.

Reply:
Hi @andreeaflorescu !

Sorry for questioning but according to the defination of kvm_device_attr from the kernel api:
```
struct kvm_device_attr {
	__u32	flags;		/* no flags currently defined */
	__u32	group;		/* device-defined */
	__u64	attr;		/* group-defined */
	__u64	addr;		/* userspace address of attr data */
};
```
I think this `addr` means the address of `nr_irqs` instead of the value of `nr_irqs`, and therefore I think we cannot simply use `addr: 128` (the test will fail too). Firecracker has a similar [implementation](https://github.com/firecracker-microvm/firecracker/blob/f55da99544f16280c31919b2596522c18d83e5e1/src/arch/src/aarch64/gic.rs#L99) of this and it also uses the pointer and similar conversion, so we probably need to keep this?

Reply:
You are right, addr is the address of nr_irqs. My bad, I was not paying attention.

We should keep it as you wrote it right now.